hudi-common/src/main/java/org/apache/hudi/common/table/read/IncrementalQueryAnalyzer.java

```diff
@@ -0,0 +1,425 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.common.table.read;
+
+import org.apache.hudi.common.model.HoodieTableType;
+import org.apache.hudi.common.table.HoodieTableMetaClient;
+import org.apache.hudi.common.table.log.InstantRange;
+import org.apache.hudi.common.table.timeline.CompletionTimeQueryView;
+import org.apache.hudi.common.table.timeline.HoodieArchivedTimeline;
+import org.apache.hudi.common.table.timeline.HoodieInstant;
+import org.apache.hudi.common.table.timeline.HoodieTimeline;
+import org.apache.hudi.common.util.ClusteringUtils;
+import org.apache.hudi.common.util.VisibleForTesting;
+import org.apache.hudi.common.util.collection.Pair;
+
+import javax.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+/**
+ * Analyzer for incremental queries.
+ *
+ * <p>The analyzer can supply info about the incremental queries including:
+ * <ul>
+ *   <li>The archived instant candidates;</li>
+ *   <li>The active instant candidates;</li>
+ *   <li>The instant filtering predicate, e.g the instant range;</li>
+ *   <li>Whether the query starts from the earliest;</li>
+ *   <li>Whether the query ends to the latest;</li>
+ *   <li>The max completion time used for fs view file slice version filtering.</li>
+ * </ul>
+ *
+ * <p><h2>Criteria for different query ranges:</h2>
+ *
+ * <table>
+ *   <tr>
+ *     <th>Query Range</th>
+ *     <th>File Handles Decoding</th>
+ *     <th>Instant Filtering Predicate</th>
+ *   </tr>
+ *   <tr>
+ *     <td>[earliest, _]</td>
+ *     <td>The latest snapshot files from table metadata</td>
+ *     <td>_</td>
+ *   </tr>
+ *   <tr>
+ *     <td>[earliest, endTime]</td>
+ *     <td>The latest snapshot files from table metadata</td>
+ *     <td>'_hoodie_commit_time' in setA, setA is a collection of all the instants completed before or on 'endTime'</td>
+ *   </tr>
+ *   <tr>
+ *     <td>[_, _]</td>
```

Comment:
should we even allow this? seems like invalid input to Incremental query. Can we simplify to just . 
````
[startTime, endTime]
[startTime, _] (we assume _ is latest time)
[_, endTime] (we assume _ is earliest time)
```

Reply:
We always assume `_` as the latest available commit, which is more matching to the streaming semantics, when both start and end time are `_`, just return the latest commit.

Reply:
When can it happen that both startTime and endTime are null/empty? 
Also, since this is a range, it looks slightly confusing as a user. I mean if startTime is null/empty then should we not treat the earliest instant as the start of the range? Just like if endTime is null/emoty then we pick latest instant as the end of the range.